### PR TITLE
dfu: mcuboot: allow only to erase secondary slot

### PR DESCRIPTION
**Pull request by mniestroj**
So far calling boot_erase_img_bank() allowed us to erase any flash area
defined in device-tree. This function should be used only for erasing
mcuboot related partitions, which are FLASH_AREA_IMAGE_PRIMARY,
FLASH_AREA_IMAGE_SECONDARY and FLASH_AREA_IMAGE_SCRATCH. Out of them
only FLASH_AREA_IMAGE_SECONDARY makes sense to be erased, because it is
the place where new firmware is going to be flashed.

Verify that FLASH_AREA_IMAGE_SECONDARY area id is used in
boot_erase_img_bank() implementation, return -EINVAL otherwise.

### CODE REVIEW

**Comment by mniestroj** [comment]
Let that PR be an entry point for discussion on how to restrict `boot_erase_img_bank()` argument value. I think that we should have common definition for those two cases:

`subsys/dfu/img_util/flash_img.c`
```
/* FLASH_AREA_ID() values used below are auto-generated by DT */
#ifdef CONFIG_TRUSTED_EXECUTION_NONSECURE
#define UPLOAD_FLASH_AREA_ID FLASH_AREA_ID(image_1_nonsecure)
#else
#if FLASH_AREA_LABEL_EXISTS(image_1)
	#define UPLOAD_FLASH_AREA_ID FLASH_AREA_ID(image_1)
#else
	#define UPLOAD_FLASH_AREA_ID FLASH_AREA_ID(image_0)
#endif
#endif /* CONFIG_TRUSTED_EXECUTION_NONSECURE */
```

`subsys/dfu/boot/mcuboot.c`
```
/* FLASH_AREA_ID() values used below are auto-generated by DT */
#ifdef CONFIG_TRUSTED_EXECUTION_NONSECURE
#define FLASH_AREA_IMAGE_PRIMARY FLASH_AREA_ID(image_0_nonsecure)
#define FLASH_AREA_IMAGE_SECONDARY FLASH_AREA_ID(image_1_nonsecure)
#else
#define FLASH_AREA_IMAGE_PRIMARY FLASH_AREA_ID(image_0)
#if FLASH_AREA_LABEL_EXISTS(image_1)
#define FLASH_AREA_IMAGE_SECONDARY FLASH_AREA_ID(image_1)
#endif
#endif /* CONFIG_TRUSTED_EXECUTION_NONSECURE */
```

Additionally there is `subsys/usb/class/usb_dfu.c` which defines:
```
#if FLASH_AREA_LABEL_EXISTS(image_1)
	#define UPLOAD_FLASH_AREA_ID FLASH_AREA_ID(image_1)
#else
	#define UPLOAD_FLASH_AREA_ID FLASH_AREA_ID(image_0)
#endif
```

Each file tries to do very similar things. We should have single common place for that.

Last, but not least, there are files which use `FLASH_AREA_ID(image_1)` directly:
- `subsys/mgmt/updatehub/updatehub.c`
- `subsys/mgmt/hawkbit/hawkbit.c`
which most likely is not working properly with `CONFIG_TRUSTED_EXECUTION_NONSECURE`.

**Comment by utzig** [comment]
This implementation might be OK for now, but soon XIP support will be functional so it will require that the app knows from which partition it is running from and only allows the complementary partition to be erased. Maybe protecting the bootloader area by only allowing primary/secondary to be erased would also be an OK solution for the moment being. What do you think? @nvlsianpu will have better input about this implementation.

**Comment by nvlsianpu** [comment]
Patch is perfectly fine for what we have supported now. direct-xip support will rework this for sure.
@de-nordic Am I right?

**Comment by de-nordic** [comment]
> 
> 
> Patch is perfectly fine for what we have supported now. direct-xip support will rework this for sure.
> @de-nordic Am I right?

I have been testing this with two overlays that have reversed roles of slots (they have not been provided here). Most of tests I have done with the slot1 being the slot where dfu sample is running and the slot0 is target for the DFU.

I do not think that there will be rework for Direct-XIP here as it is just the matter of switching the `zephyr,code-partition` with `zephyr,dfu-code-partition` .

I have faced some problem though, but I need some more testing to be sure, namely: when I have been running the dfu sample from slot1 and did the dfu to slot-0 it worked fine; then I did the dfu again to slot-1 but the device would boot to slot-0 nevertheless; I had to damage slot-0 to make the device boot to slot-1 again.

**Comment by de-nordic** [comment]
> 
> 
> Patch is perfectly fine for what we have supported now. direct-xip support will rework this for sure.
> @de-nordic Am I right?

Yes. I am changing definitions of slots and this will be changed too. Direct-XIP allows code to be executed from either slot, which means that an application built for slot-1 would still be able to write over itself.

There is also possibility to relocate code into ram and execute it from there; such code should be able to erase both slots.

I am not sure whether this is needed. This is something that should be tested by user, so that the slot you are running your software from is not erased.
Basically what would happen in normal situation:  you get this error which means that your software is somehow broken and tries to overwrite itself; you fix that; you newer get the error again.

**Comment by de-nordic** [comment]
> Last, but not least, there are files which use `FLASH_AREA_ID(image_1)` directly:
> 
>     * `subsys/mgmt/updatehub/updatehub.c`
> 
>     * `subsys/mgmt/hawkbit/hawkbit.c`
> 
>     * mcumgr's `cmd/img_mgmt/port/zephyr/src/zephyr_img_mgmt.c`
> 
> 
> which most likely is not working properly with `CONFIG_TRUSTED_EXECUTION_NONSECURE`.

I am working on moving the defines for "FLASH_AREAS" into common header; additionally, while working on Direct-XIP support, the selection of running/dfu image has been moved to DTS (https://github.com/zephyrproject-rtos/zephyr/pull/29974).

**Comment by mniestroj** [comment]
So what is the conclusion here about restricting `boot_erase_img_bank()` arguments? As you can see unit tests are failing, because `boot_erase_img_bank(FLASH_AREA_ID(image_0)` is executed there.

Maybe we should just change to `if (area_id != FLASH_AREA_IMAGE_PRIMARY && area_id != FLASH_AREA_IMAGE_SECONDARY) {`, as @utzig suggested. We can always improve this in future, but at least bootloader will be protected and we will not touch any other flash areas, such as settings storage, etc. I also guess that situation will improve when @de-nordic will cleanup "FLASH_AREAS" macros.

**Comment by de-nordic** [comment]
> So what is the conclusion here about restricting `boot_erase_img_bank()` arguments? As you can see unit tests are failing, because `boot_erase_img_bank(FLASH_AREA_ID(image_0)` is executed there.

The `subsys/dfu/mcuboot/src/main.c:107` error happens because the test tests operations on `slot_0` and the commit forbids such operations. The test needs to be altered for the commit.

> Maybe we should just change to `if (area_id != FLASH_AREA_IMAGE_PRIMARY && area_id != FLASH_AREA_IMAGE_SECONDARY) {`, as @utzig suggested.

Or, the test could be done against the partitions you do not want to overwrite, only allow writes to partitions that are not `boot0` and `zephyr,code-partition`. This would prevent code from overwriting bootloader and running application.

You may also add Kconfig options that would enable/disable such protections when desired by the programmer; this way ,when someone wants to write anywhere, there would be no need for dev to alter code just select Kcofnig options instead.

**Comment by nvlsianpu** [comment]
@mniestroj #30370 was merged. Shall we close this PR then?

**Comment by mniestroj** [comment]
@nvlsianpu This PR is not strictly related to #30370, however I do not plan to do any changes to this PR. Besides, @de-nordic wanted to move "FLASH_AREAS" to common headers, so it should be easier to decide which partition is "safe to be erased" after that cleanup. Closing this PR for now.
